doc/designs/quic-design/quic-connection-id-cache.md

```diff
@@ -0,0 +1,204 @@
+objects.  This will be changed with Path migration support and QUIC
+server support._
+
+API
+---
+
+```c
+struct quic_conn_id_cache_st;
+
+typedef struct quic_conn_id_cache_st QUIC_CONN_ID_CACHE;
+
+/*
+ * Allocate and free a connection ID cache
+ */
+QUIC_CONN_ID_CACHE *ossl_quic_new_conn_id_cache(void);
+void ossl_quic_free_conn_id_cache(QUIC_CONN_ID_CACHE *cache);
+
+/*
+ * Add an ID to connection mapping to the cache.
+ * One call is to add our local ID, the other to add our peer's remote ID.
+ */
+int ossl_quic_set1_local_conn_id(QUIC_CONN_ID_CACHE *cache,
+                                 const SSL_CONNECTION *conn,
+                                 const unsigned char *conn_id,
+                                 size_t conn_id_len);
+int ossl_quic_set1_remote_conn_id(QUIC_CONN_ID_CACHE *cache,
+                                  const SSL_CONNECTION *conn,
+                                  const unsigned char *conn_id,
+                                  size_t conn_id_len);
+
+/*
+ * Remove an ID to connection mapping to the cache.
+ * One call is to remove our local ID, the other to remove our peer's remote ID.
+ */
+int ossl_quic_remove_local_conn_id(QUIC_CONN_ID_CACHE *cache,
+                                   const unsigned char *conn_id,
+                                   size_t conn_id_len);
+int ossl_quic_remove_remote_conn_id(QUIC_CONN_ID_CACHE *cache,
+                                    const unsigned char *conn_id,
+                                    size_t conn_id_len);
+
+/*
+ * Remove a connection and all associated cached IDs
+ */
+int ossl_quic_remove_connection(QUIC_CONN_ID_CACHE *cache,
+                                const SSL_CONNECTION *conn);
+
+/*
+ * Lookup a connection by ID.
+ * One call is to lookup based on a local ID, the other to lookup based our
+ * peer's remote ID.
+ */
+SSL_CONNECTION *ossl_quic_get0_local_conn_id(const QUIC_CONN_ID_CACHE *cache,
+                                             const unsigned char *conn_id,
+                                             size_t conn_id_len);
+SSL_CONNECTION *ossl_quic_get0_remote_conn_id(const QUIC_CONN_ID_CACHE *cache,
+                                              const unsigned char *conn_id,
+                                              size_t conn_id_len);
+
+/*
+ * Query the most recent ID for a connection.
+ * One call is to query the most recent local ID, the other to query the most
+ * recent remote ID.
+ *
+ * These would better be stored in the connection itself.
+ */
+int ossl_get0_local_conn_id(const QUIC_CONN_ID_CACHE *cache,
+                            const SSL_CONNECTION *conn,
+                            unsigned char **conn_id,
+                            size_t *conn_id_len);
+int ossl_get0_remote_conn_id(const QUIC_CONN_ID_CACHE *cache,
+                             const SSL_CONNECTION *conn,
+                             unsigned char **conn_id,
+                             size_t *conn_id_len);
```

Comment:
These two functions need more discussion of the conn_id_len parameter.

You could:
- Check `*conn_id_len` is big enough on call and write it on return with the actual size, orr
- Specify the buffer must always be at least 20 bytes long (since connection IDs cannot be bigger than this), and justt write it on return with the actual size.

This might be a good opportunity to also add a `#define` for the maximum connection ID length.

Reply:
Does the cache need to define/enforce the limits?  Seems kind of out of place.  The RFC says 1 - 20 bytes however, it notes in 17.2 that implementations SHOULD accept longer IDs from later QUIC versions (i.e. there is no maximum).

For MVP it would be possible to switch to arrays for the conn_id to save mallocs (as an optimisation).  If I do, there will need to be a define for the maximum.  For the more general case, I'd need to think it though.  It might be reasonable but 17.2.

Reply:
Later versions might increase the maximum connection ID length but I'd be surprised if that happens anytime soon. There's a compromise regarding connection ID length since the connection ID has to be in every packet header. 20 bytes is already quite long.

Personally I'd go with a fixed limit of 20 bytes and use a fixed-length array. This is what I'm doing myself in any case. We can change it later in the (IMO unlikely) event long connection IDs become a thing, it seems overkill for now.